### DISCUSSION TITLE
[terraform]  Problem Adding 10 or more streams to connection

### DISCUSSION THREAD

**Discussion by corbin-Tprime** [airbytehq/airbyte#31472] [General]
Hi All,

I was working on creating a new connection that uses a redshift source and an s3 destination. I was able to create the source and destinations just fine with Airbyte/terraform. However, when trying to create the connection, the API seems to fail when trying to add more than 10 streams to the configuration.

The error I receive:
```
Error: failure to invoke API
│ 
│   with airbyte_connection.redacted_name,
│   on connections.tf line 5, in resource "airbyte_connection" "redacted_name":
│    5: resource "airbyte_connection" "redacted_name" {
│ 
│ error sending request: Patch "https://airbyte.redactedname.com/v1/connections/redacted-id": context deadline exceeded (Client.Timeout
│ exceeded while awaiting headers)
``` 

I did redact some of the information in the error message due to the nature of the work. That said, it shouldn't be relevant to the issue as creating the connection does work. We just can't add more than 10 streams to the configuration.  Once we try to add an 11th+, we receive the API error listed above.

Thanks!

**Comment by corbin-Tprime** [comment]
@marcosmarxm Any idea on a potential work around or if this has been resolved?

**Reply by marcosmarxm** [reply]
Hello @corbin-Tprime sorry the missing reply here. @pgrant87 do you mind trying replicate the issue and escalate to engineer if needed?

**Reply by pgrant87** [reply]
Sure thing, let me try to replicate this @corbin-Tprime and I'll get back to you with my findings

**Reply by corbin-Tprime** [reply]
@marcosmarxm @pgrant87 Thank you! If you are not able to replicate, any ideas on how to troubleshoot or a work around would be awesome! I did see another user had a similar issue/discussion: https://github.com/airbytehq/airbyte/discussions/31557

From what I can tell, there isn't a setting for the API to keep the connection alive, so Im a bit at a loss here, and with connections that have 30+ streams being able to enable with code will save countless hours of work for us!

-Corbin

**Reply by pgrant87** [reply]
No problem @corbin-Tprime i'll let you know how it all goes! Thanks for bringing 31557 to our attention, that does seem like a related issue.

Just to get some additional context here, would you be able to tell me what connectors are involved here and their versions? What version of Airbyte are you running and finally would you be able to share your connection Terraform code? 

Thank you

**Reply by corbin-Tprime** [reply]
@pgrant87 sure, unfortunately, this doesnt seem to be limited to the connector.
### Airbyte Version
`0.50.33`
### Source Connectors
```
Snowflake 0.2.2
Postgres 3.2.21
```
### Destination Connectors
```
Postgres 0.4.0
S3 0.5.1
Snowflake 3.4.10
```
Redacted Terraform code(I cant share the naming conventions we use, but the code is below:
```
resource "airbyte_connection" "name" {
    name = "connection_name"
    destination_id = destination_id
    source_id = source_id
    status = "active"
    data_residency = "auto"
    namespace_definition = "custom_format"
    namespace_format = "namespace_name"
    schedule = {
      cron_expression = "0 0 5 * * ? UTC"
      schedule_type   = "cron"
    }
    configurations = {
      streams = [ {
        sync_mode    = "incremental_deduped_history"
        cursor_field = ["UPDATED_AT"]
        primary_key = [[ "ID" ]]
        name         = "table8"               
      },
      {
        sync_mode    = "full_refresh_overwrite"
        name         = "table7"               
      },
      {
        sync_mode    = "incremental_deduped_history"
        cursor_field = ["UPDATED_AT"]
        primary_key = [[ "ID" ]]
        name         = "table6"               
      },
      {
        sync_mode    = "incremental_deduped_history"
        cursor_field = ["UPDATED_AT"]
        primary_key = [[ "ID" ]]
        name         = "table9"               
      },
      {
        sync_mode    = "incremental_deduped_history"
        cursor_field = ["UPDATED_AT"]
        primary_key = [[ "ID" ]]
        name         = "table10"               
      },
      {
        sync_mode    = "incremental_deduped_history"
        cursor_field = ["UPDATED_AT"]
        primary_key = [[ "ID" ]]
        name         = "table"               
      },
      {
        sync_mode    = "incremental_deduped_history"
        cursor_field = ["UPDATED_AT"]
        primary_key = [[ "ID" ]]
        name         = "table2"               
      },
      {
        sync_mode    = "incremental_deduped_history"
        cursor_field = ["UPDATED_AT"]
        primary_key = [[ "ID" ]]
        name         = "table3"               
      },
      {
        sync_mode    = "incremental_deduped_history"
        cursor_field = ["UPDATED_AT"]
        primary_key = [[ "ID" ]]
        name         = "table4"               
      },
      {
        sync_mode    = "incremental_deduped_history"
        cursor_field = ["UPDATED_AT"]
        primary_key = [[ "ID" ]]
        name         = "table5"               
      },
       ]
    }
}
```

The error seems to be a timeout error on the server side.

**Reply by corbin-Tprime** [reply]
@pgrant87 Any update here or estimate on when you can have verification results?

**Reply by pgrant87** [reply]
hi @corbin-Tprime thanks for those additional details, I appreciate it!

Just wanted to give you an update on how my testing went today. I've tried hitting the create a connection API endpoint while passing a few streams but I haven't been able to replicate yet. I started with 5 streams, incrementing by 5 streams for each test, up to a total of 50 but each time the connection was created. Tomorrow I'm going to drastically increase the stream number to see if this will cause the timeout. I used our E2E testing source so I might try a different source as well but I'll let you know how this goes

**Reply by corbin-Tprime** [reply]
@pgrant87 Could you clarify what the E2E testing source is? For some clarity on our Airbyte implementation, we host on EC2 instances that have 8-16 cores. Is there some sort of setting that I can adjust when interacting the API? Or is there a setting we need to adjust with Terraform?

Following up, on the above:
are you hitting the Airbyte API or using the terraform airbyte_connection resource? 
Are you behind a VPN when trying to hit the API?

**Reply by pgrant87** [reply]
Thanks again for the additional context, docs on the E2E testing source are located here https://docs.airbyte.com/integrations/sources/e2e-test

For my testing, I'm hitting the API directly as this is what the terraform provider uses. I'm not behind a VPN. I tried upping the streams count today drastically but still not having any luck hitting the timeout. I'll continue testing on Monday and investigate to see if there's a setting to increase this time out in the meantime

**Reply by corbin-Tprime** [reply]
@pgrant87 we were using `v 0.3.5` terrafrom provider, after upgrading to `v 0.3.7` we seem to be able to add the streams. So maybe the issue was addressed in the version change

**Reply by corbin-Tprime** [reply]
@pgrant87 Actually, I need to retract my previous statement. When using apply, the connection eventually shows up in airbyte, however, the terraform provider still produces the `Client.Timeout exceeded while awaiting headers" error and the state of the connection is not stored in our `terraform state`.

And while the API is used with terraform, couldn't there be an issue with the airbyte terraform code that causes this issue?
You probably want to test the terraform provider and not just hit the API directly to see what I am seeing?

**Reply by pgrant87** [reply]
Just to confirm, you're now receiving the `Client.Timeout exceeded while awaiting headers` error but the connection does get created successfully within the UI?

There could potentially be something wrong with the terraform provider but as it's a timeout error I think it's more likely to be an issue with the API. I will try testing via Terraform code to be sure and I'll let you know how that goes.

One final question, is your installation on kubernetes or docker-compose? Was this a fresh install at 0.50.33 or did you upgrade to this version? The reason I ask, there were some fixes around API timeouts included in 0.50.31 that may need a slight config change

**Reply by corbin-Tprime** [reply]
@pgrant87 Yea, after upgrading the terraform provider, we are able to create connections, however, the timeout message still persists.  

We upgraded from a previous version of Airbyte to 0.50.33. What are these config changes?

**Reply by pgrant87** [reply]
The changes can be found here, https://github.com/airbytehq/airbyte/issues/29506 This is a similar issue to yours but not exactly the same error. A PR was put up for this, comment on Oct 20:

> That release both includes some increased default timeouts as well as a change to how our micronaut server allocates threads in order to avoid timing out.

The main change to the settings is `http://` was added to `INTERNAL_API_HOST` either in .env for docker-compose installs or the env config map for kubernetes/helm. I noticed this tripped a few users up

I have also tried to replicate further this morning using terraform to set up the connection but again mine succeded without error. Can I ask, how long does it take before your `terraform apply` errors? Currently my requests take about 15 seconds to process but they get applied with no issue

**Reply by corbin-Tprime** [reply]
@pgrant87 When I try to add `http://` in the `INTERNAL_API_HOST` variable, I get a 502 gateway error when booting my containers back up.  What version of airbyte do I need to be on for the above mentioned fix to work?

In terms of timeout, the apply attempts to apply for 60 seconds then produces the timeout error. The connection does get created, but the state is not stored so i have to go in and then import the state from the created connection.

**Reply by pgrant87** [reply]
That fix should be in 0.50.31 onwards. Would it be possible to attach the file where you'd added the `http://` with any sensitive info redacted so I can take a look?

**Reply by corbin-Tprime** [reply]
@pgrant87 We just have the out of box `.env` file for version `0.50.33`, we use param store to pass in most variables.

Here is the file: 
[env-file.txt](https://github.com/airbytehq/airbyte/files/13663804/env-file.txt)

**Reply by pgrant87** [reply]
Thanks for passing that on, your `.env` looks absolutely fine, nothing out of the ordinary. When you received the 502 gateway error when you initially added the `http://` to INTERNAL_API_HOST, had all containers started without error or did you notice any other issues when spinning up?

On the Terraform side of things, can I ask, do you have many other large resources? My testing simply consisted of a source, destination and a connection with lots of streams being applied but they were always dealt with within about 20 seconds

[a comment by an automated account is not shown]
